lib/plottr_components/src/components/project/BookDialog.js

```diff
@@ -24,38 +30,67 @@ const BookDialogConnector = (connector) => {
       this.themeRef = null
     }
 
-    saveEdit = () => {
-      const { book } = this.props
+    saveEdit = (saveAndOpen) => {
+      const { book, books } = this.props
 
       let title = this.titleRef.value
       let premise = this.premiseRef.value
       let genre = this.genreRef.value
       let theme = this.themeRef.value
-      this.props.actions.editBook(book.id, { title, premise, genre, theme })
-      this.props.cancel()
+      const newBookId = objectId(books.allIds)
+
+      if (!book) {
+        this.props.actions.addBook(newBookId, { title, premise, genre, theme })
+      } else {
+        this.props.actions.editBook(book.id, { title, premise, genre, theme })
+      }
+      this.handleCancel()
+
+      if (saveAndOpen) {
+        this.navigateToBook(book?.id || newBookId)
+      }
+    }
+
+    handleCancel = () => {
+      this.props.uiActions.setBookDialogClose()
+    }
+
+    navigateToBook = (bookId) => {
+      this.props.uiActions.navigateToBookTimeline(bookId, inBrowser, browserHistory)
     }
 
     renderToolBar() {
       return (
         <ButtonToolbar>
-          <Button bsStyle="success" onClick={this.saveEdit}>
+          <Button onClick={this.handleCancel}>{i18n('Cancel')}</Button>
+          <Button bsStyle="success" onClick={() => this.saveEdit()}>
             {i18n('Save')}
           </Button>
-          <Button onClick={this.props.cancel}>{i18n('Cancel')}</Button>
+          <Button className="pull-right" bsStyle="primary" onClick={() => this.saveEdit(true)}>
+            {i18n('Save and Open')}
+          </Button>
         </ButtonToolbar>
       )
     }
 
+    handleDown = (event) => {
+      if (event.key === 'Enter') this.saveEdit()
+    }
+
     renderBody() {
-      const { book } = this.props
+      const { book, bookNumber, bookDialogBookNumber } = this.props
       return (
-        <Form horizontal>
+        <Form horizontal onKeyDown={this.handleDown}>
           <FormGroup>
             <Col componentClass={ControlLabel} sm={3}>
               {i18n('Book #')}
             </Col>
             <Col sm={8}>
-              <span className="lead">{this.props.bookNumber}</span>
+              {/* bookNumber will be fetched from the current `book`
+                if `book` is not empty (editBook mode)
+                else bookDialogBookNumber (addBook mode)
+              */}
+              <span className="lead">{bookNumber || bookDialogBookNumber}</span>
```

Comment:
Please write a `bookNumberSelector` to replace the bookDialogBookNumberSelector` and remove `bookDialogBookNumber`.
The selector should produce the number of books in the project when there's no `bookId` for this dialog.